File: Test/src/main/scala/org/example/Top_Rate_Redis_Hbase.scala
package org.example

import com.google.gson.Gson
import org.apache.flink.api.common.eventtime.WatermarkStrategy
import org.apache.flink.api.common.state.{ListState, ListStateDescriptor}
import org.apache.flink.configuration.Configuration
import org.apache.flink.connector.kafka.source.KafkaSource
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.functions.sink.{RichSinkFunction, SinkFunction}
import org.apache.flink.streaming.api.scala.function.WindowFunction
import org.apache.flink.streaming.api.scala.{OutputTag, StreamExecutionEnvironment, createTypeInformation}
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.streaming.connectors.redis.common.mapper.{RedisCommand, RedisCommandDescription, RedisMapper}
import org.apache.flink.streaming.util.serialization.SimpleStringSchema
import org.apache.flink.util.Collector
import org.apache.hadoop.hbase._
import org.apache.hadoop.hbase.client._

import scala.jdk.CollectionConverters.{iterableAsScalaIterableConverter, seqAsJavaListConverter}
import scala.util.Random

object Top_Rate_Redis_Hbase {
  private val Bigdata1 = "192.168.45.16"
  private val BootstrapServers = s"$Bigdata1:9092"
  private val topic = "ods_mall_data"
  val MaxOutOfOrderness: Long = 5 * 1000L
  // "已发货", "已付款", "已下单","已签收",
  val tuple = Array("已发货", "已签收")
  val canceltag = new OutputTag[Order_master]("canceltag")

  def main(args: Array[String]): Unit = {
    val handler = FlinkHandlerHelper.flinkHandlerHelper()
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    val kafkaSource = KafkaSource.builder[String]
      .setBootstrapServers(BootstrapServers)
      .setTopics(topic)
      .setStartingOffsets(OffsetsInitializer.earliest())
      .setValueOnlyDeserializer(new SimpleStringSchema())
      .build()
    val orderData = env.fromSource(kafkaSource, WatermarkStrategy.noWatermarks(), topic)
      .map(in => {
        val gson = new Gson
        val ods_mall_data: Ods_mall_data = gson.fromJson(in, classOf[Ods_mall_data])
        ods_mall_data
      })
      .filter(_.table.contains("order_master"))
      .map(in => {
        val gson = new Gson
        val order_master: Order_master = gson.fromJson(gson.toJson(in.data), classOf[Order_master])
        order_master
      })
    orderData.print()
    //      // 设置EventTime
    //      .assignTimestampsAndWatermarks(new choseEventTime)
    //      .keyBy(_ => true)
    //      .window(TumblingEventTimeWindows.of(Time.minutes(1)))
    //      .trigger(EventTimeTrigger.create())
    //      .apply(new putEventData)

    //    val dwdData = orderData.keyBy(_ => true)
    //      .process(new myPF)
    val cancel = orderData

//      .map(i => {
//                val num: Int = Random.nextInt(10)
//                val name = new getAttributeName
//                val listName: Array[String] = name.getClassFields(classOf[Order_master])
//                val value: Array[String] = i.productIterator.map(_.toString).toArray
//                val tuples: Array[(String, String, String)] = listName.zip(value).map { case (l, v) => ("info", l, v) }
//                (s"$num${i.create_time}", tuples)
//              })
//      .addSink(handler.hbaseWrite("test:qwe"))
    //    val cancel = dwdData
    //      .getSideOutput(canceltag)
    //    val cancelrate = cancel
    //      .map(_ => ("cancelrate", 1.0))
    //      .union(orderData.map(_ => ("all", 1.0)))
    //      .keyBy(_ => true)
    //      .window(TumblingEventTimeWindows.of(Time.minutes(1)))
    //      .apply(new winFunction)

    //    val config = new FlinkJedisPoolConfig.Builder()
    //      .setHost(Bigdata1)
    //      .setPort(6379)
    //      .build()

    //1      要在Redis get到中文 redis-cli --raw
    //    dwdData.addSink(new RedisSink[(String, String)](config, new redisSink))

    //2
    //    cancelrate.addSink(new RedisSink[(String, String)](config, new redisSink))
    //3     在hbase显示中文 scan 'test:order_master', {FORMATTER => 'toString',COLUMNS => 'info:shipping_user',LIMIT => 5,REVERSED => true}
//    cancel.addSink(new writeHbase(Bigdata1, "test", "order_master", "info"))
    env.execute()
  }

  class putEventData extends WindowFunction[Order_master, Order_master, Boolean, TimeWindow] {
    override def apply(key: Boolean, window: TimeWindow, input: Iterable[Order_master], out: Collector[Order_master]): Unit = {
      // 对窗口内的数据按时间戳进行排序
      val sortedInput = input.toList
      // 发出按时间顺序排序后的数据
      sortedInput.foreach { case (in) =>
        out.collect(in)
      }
    }
  }

  class winFunction extends WindowFunction[(String, Double), (String, String), Boolean, TimeWindow] {
    override def apply(key: Boolean, window: TimeWindow, input: Iterable[(String, Double)], out: Collector[(String, String)]): Unit = {
      val str = input.toList.groupBy(_._1).mapValues(_.map(_._2).sum)
      val cancel = str.getOrElse("cancelrate", 0.0)
      val all = str.getOrElse("all", 0.0)
      val value = ((cancel / all) * 1000).round / 10.0
      out.collect("cancelrate", s"$value%")
    }
  }

  class myPF extends ProcessFunction[Order_master, (String, String)] {
    lazy val top2UserConsumption: ListState[((Long, String), Double)] = getRuntimeContext.getListState(new ListStateDescriptor[((Long, String), Double)]("top2UserConsumption", classOf[((Long, String), Double)]))

    override def processElement(i: Order_master, context: ProcessFunction[Order_master, (String, String)]#Context, collector: Collector[(String, String)]): Unit = {
      if (tuple.contains(i.order_status)) {
        top2UserConsumption.add((i.order_id, i.shipping_user), i.payment_money)
        val longToTuples = top2UserConsumption.get().asScala.groupBy(_._1)
          .mapValues(_.map(_._2).sum)
          .toList.sortBy(_._2)
          .reverse
        top2UserConsumption.clear()
        top2UserConsumption.addAll(longToTuples.asJava)
        val put = longToTuples.map(d => s"${d._1._1}:${d._1._2}:${d._2}") :+ "null"
        collector.collect("top2userconsumption", s"[${put(0)},${put(1)}]")

      }
      else if (i.order_status.contains("已付款"))
        context.output(canceltag, i)
    }
  }

  class redisSink extends RedisMapper[(String, String)] {
    override def getCommandDescription: RedisCommandDescription = new RedisCommandDescription(RedisCommand.SET)

    override def getKeyFromData(t: (String, String)): String = t._1

    override def getValueFromData(t: (String, String)): String = t._2
  }

  private class writeHbase(zookeeper: String, namepace: String, table: String, family: String) extends RichSinkFunction[Order_master] {
    private var connection: Connection = _
    private var htable: Table = _

    override def open(parameters: Configuration): Unit = {
      val hbaseConf = HBaseConfiguration.create()
      hbaseConf.set(HConstants.ZOOKEEPER_QUORUM, zookeeper)
      hbaseConf.set(HConstants.ZOOKEEPER_CLIENT_PORT, "2181")
      connection = ConnectionFactory.createConnection(hbaseConf)
      htable = connection.getTable(TableName.valueOf(namepace, table))
    }

    override def close(): Unit = {
      if (connection != null) {
        connection.close()
      }
      if (htable != null) {
        htable.close()
      }
    }

    override def invoke(value: Order_master, context: SinkFunction.Context): Unit = {
      val num: Int = Random.nextInt(10)
      val name = new getAttributeName
      val listName: Array[String] = name.getClassFields(classOf[Order_master])
      val values: Array[String] = value.productIterator.map(_.toString).toArray
      val tuples: Array[(String, String, String)] = listName.zip(values).map { case (l, v) => ("info", l, v) }
      val rowkey = s"$num${value.create_time}"
      val put = new Put(rowkey.getBytes())
      for (i <- tuples) {
        put.addColumn(i._1.getBytes, i._2.getBytes, i._3.getBytes)
      }
      htable.put(put)
    }
  }
}

File: Task/src/main/java/org/example/gs3/TopRateRedisHbase.java
package org.example.gs3;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.*;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;
import org.example.*;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Random;

public class TopRateRedisHbase implements Constants {
    static String topic = "ods_mall_data";

    public static void main(String[] args) throws Exception {
        OutputTag<OrderMaster> cancel = new OutputTag<OrderMaster>("cancel") {
        };
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.setParallelism(1);

        KafkaSource<String> source = new KafkaConnect().kafkaReader(topic);
        WatermarkStrategy<OrderMaster> orderMasterWatermarkStrategy = new Watermarks().orderMasterWatermarkStrategy();

        ProcessFunction<OrderMaster, String> processFunction = new ProcessFunction<OrderMaster, String>() {
            @Override
            public void processElement(OrderMaster orderMaster, ProcessFunction<OrderMaster, String>.Context context, Collector<String> collector) throws Exception {
                if (Arrays.asList(filter).contains(orderMaster.getOrder_status())) {
                    TopN topN = new TopN();
                    String member = orderMaster.getCustomer_id() + ":" + orderMaster.getShipping_user();
                    topN.addElem("userconsumption", orderMaster.getPayment_money(), member);
                    Tuple2<String, Double>[] userconsumptions = topN.getTopN("userconsumption", 0, 2, true);
                    String put = "[" + userconsumptions[0].f0 + ":" + userconsumptions[0].f1 + "," + userconsumptions[1].f0 + ":" + userconsumptions[1].f1 + "]";
                    collector.collect(put);
                } else if (orderMaster.getOrder_status().equals("已退款")) context.output(cancel, orderMaster);
            }
        };

        SingleOutputStreamOperator<OdsMallData> odsMallDataStream = env
//                .fromSource(source, WatermarkStrategy.noWatermarks(), topic)
                .readTextFile("E:\\input\\order_master.txt")
                .map(i -> {
                    Gson gson = new GsonBuilder().create();
                    OdsMallData odsMallData = gson.fromJson(i, OdsMallData.class);
                    return odsMallData;
                });

        SingleOutputStreamOperator<OrderMaster> order_master = odsMallDataStream.filter(i -> i.getTable().equals("order_master"))
                .map(i -> {
                    Gson gson = new GsonBuilder().create();
                    String json = gson.toJson(i.getData());
                    OrderMaster orderMaster = gson.fromJson(json, OrderMaster.class);
                    return orderMaster;
                }).assignTimestampsAndWatermarks(orderMasterWatermarkStrategy);
        SingleOutputStreamOperator<String> top2user = order_master
                .keyBy(i -> true)
                .process(processFunction);

        SingleOutputStreamOperator<Tuple2<String, Integer>> orderCounts = order_master
                .map(new MapFunction<OrderMaster, Tuple2<String, Integer>>() {
                    @Override
                    public Tuple2<String, Integer> map(OrderMaster orderMaster) {
                        return new Tuple2<>("orderCount", 1);
                    }
                })
                .keyBy(0)
                .window(TumblingEventTimeWindows.of(Time.minutes(1)))
                .sum(1);

        DataStream<OrderMaster> cancelOrders = top2user.getSideOutput(cancel);

        SingleOutputStreamOperator<Tuple2<String, Integer>> cancelCounts = cancelOrders
                .map(new MapFunction<OrderMaster, Tuple2<String, Integer>>() {
                    @Override
                    public Tuple2<String, Integer> map(OrderMaster orderMaster) {
                        return new Tuple2<>("cancelCount", 1);
                    }
                })
                .keyBy(0)
                .window(TumblingEventTimeWindows.of(Time.minutes(1)))
                .sum(1);

        SingleOutputStreamOperator<String> cancelRate = (SingleOutputStreamOperator<String>) orderCounts
                .join(cancelCounts)
                .where(i -> true)
                .equalTo(i -> true)
                .window(TumblingEventTimeWindows.of(Time.minutes(1)))
                .apply((cancelCount, orderCount) -> {
                    double rate = ((cancelCount.f1 * 1.0) / (orderCount.f1 * 1.0));
                    String s = String.format("%.1f", rate);
                    return s + "%";
                });
        SingleOutputStreamOperator<Tuple2<String, Tuple3<String, String, String>[]>> cancelHbase = cancelOrders
                .map(i -> {
                    int num = new Random().nextInt(10);
                    Field[] fields = OrderMaster.class.getDeclaredFields();
                    String[] listName = new String[fields.length];
                    String[] value = new String[fields.length];
                    for (int j = 0; j < fields.length; j++) {
                        fields[j].setAccessible(true);
                            listName[j] = fields[j].getName();
                            value[j] = fields[j].get(i).toString();
                    }
                    Tuple3<String, String, String>[] tuples = new Tuple3[listName.length];
                    for (int j = 0; j < fields.length; j++) {
                        tuples[j] = new Tuple3<>("info", listName[j], value[j]);
                    }
                    return new Tuple2<>(num + i.getCreate_time(), tuples);
                });
        top2user.addSink(new RedisWriter().redisSet("top2userconsumption"));
        cancelRate.addSink(new RedisWriter().redisSet("cancelrate"));
        cancelHbase.addSink(new HbaseWriter().hbaseWriter("shtd_result:order_info"));
        env.execute("TopRateRedisHbase");
    }

}